internal/discover/graphics.go

```diff
@@ -36,20 +36,20 @@ import (
 // TODO: The logic for creating DRM devices should be consolidated between this
 // and the logic for generating CDI specs for a single device. This is only used
 // when applying OCI spec modifications to an incoming spec in "legacy" mode.
-func NewDRMNodesDiscoverer(logger logger.Interface, devices image.VisibleDevices, devRoot string, nvidiaCTKPath string) (Discover, error) {
+func NewDRMNodesDiscoverer(logger logger.Interface, devices image.VisibleDevices, devRoot string, nvidiaCDIHookPath string) (Discover, error) {
```

Comment:
Using this as an example, I think we want to introduce an abstraction around a CDI hook that we inject here instead. That can contain the logic for generating the command / hook for a set of arguments.

We already have this logic in `CreateNvidiaCDIHook` (`internal/discover/hooks.go`) but it could be better organized.

Something like:
```
type HookCreator interface {
    CreateHook(name string, args ...string) discover.Hook
}
```

Where we could implement `CreateHook` as something like:
```
func (c *creator) CreateHook(name string, args ...string) Hook {
	return Hook{
		Lifecycle: cdi.CreateContainerHook,
		Path: c.Path,
		Args: append(c.requiredArgs(name), args...),
	}
}

func (c *creator) requiredArgs(name string) []string {
	base := filepath.Base(c.Path)
	if base == "nvidia-ctk" {
		return []string{base, "hook", name}
	}
	return []string{base, name}
}
```

Reply:
This would allow us to make a decision on how to construct this when we are initialising things for the `nvidia-container-runtime` or for CDI spec generation instead of passing paths down and making this decision later.

Note that this can be a seprate PR which captures the existing behaviour and then we can add the new type of hook creator as a follow-up.

Reply:
> Note that this can be a seprate PR which captures the existing behaviour and then we can add the new type of hook creator as a follow-up

That sounds like a good idea. I don't know that we want to mix these two distinct PRs: one that splits the functionality out into a library surface and creates a new CLI, with one that restructures some of how the things are called inside that "library".

And I am not 100% sure I understand it anyways, so it probably is better to have a dedicated discussion to it.

Reply:
I had a quick look at this, and although I think it would add value to better encapsulate hook creation, there are some other things that would need to be cleaned up first.

Let's ignore this from the point of view of this PR.

Note that the logic to optionally inject `"hook"` when generating the CDI spec may still be useful -- although it doesn't need the more agressive refactoring proposed here.